src/GraphBLAS-sharp.Backend/Vector/Sparse/Map2.fs

```diff
@@ -103,6 +103,102 @@ module internal Map2 =
+
+                    match (%op) leftValue rightValue with
+                    | Some value ->
+                        resultValues.[gid] <- value
+                        resultIndices.[gid] <- i
+
+                        resultBitmap.[gid] <- 1
+                    | None -> resultBitmap.[gid] <- 0 @>
+
+        let kernel =
+            clContext.Compile <| preparePositions opAdd
+
+        fun (processor: MailboxProcessor<_>) (vectorLenght: int) (leftValues: ClArray<'a>) (leftIndices: ClArray<int>) (rightValues: ClArray<'b option>) ->
+
+            let resultBitmap =
+                clContext.CreateClArrayWithSpecificAllocationMode<int>(DeviceOnly, vectorLenght)
+
+            let resultIndices =
+                clContext.CreateClArrayWithSpecificAllocationMode<int>(DeviceOnly, vectorLenght)
+
+            let resultValues =
+                clContext.CreateClArrayWithSpecificAllocationMode<'c>(DeviceOnly, vectorLenght)
+
+            let ndRange =
+                Range1D.CreateValid(vectorLenght, workGroupSize)
+
+            let kernel = kernel.GetKernel()
+
+            processor.Post(
+                Msg.MsgSetArguments
+                    (fun () ->
+                        kernel.KernelFunc
+                            ndRange
+                            vectorLenght
+                            leftValues
+                            leftIndices
+                            rightValues
+                            resultBitmap
+                            resultValues
+                            resultIndices)
+            )
+
+            processor.Post(Msg.CreateRunMsg<_, _> kernel)
+
+            resultBitmap, resultValues, resultIndices
+
+    let runSparseDense<'a, 'b, 'c when 'a: struct and 'b: struct and 'c: struct>
+        op
+        (clContext: ClContext)
+        workGroupSize
+        =
+
+        let prepare =
+            preparePositionsSparseDense<'a, 'b, 'c> clContext workGroupSize op
+
+        let setPositions =
+            Common.setPositionsOption clContext workGroupSize
+
+        fun (processor: MailboxProcessor<_>) allocationMode (leftVector: ClVector.Sparse<'a>) (rightVector: ClArray<'b option>) ->
+
+            let bitmap, allValues, allIndices =
+                prepare processor leftVector.NNZ leftVector.Values leftVector.Indices rightVector
+
+            match setPositions processor allocationMode allValues allIndices bitmap with
+            | Some (resultValues, resultIndices) ->
+
+                processor.Post(Msg.CreateFreeMsg<_>(allIndices))
+                processor.Post(Msg.CreateFreeMsg<_>(allValues))
+                processor.Post(Msg.CreateFreeMsg<_>(bitmap))
+
+                Some(
+                    { Context = clContext
+                      Values = resultValues
+                      Indices = resultIndices
+                      Size = leftVector.Size }
+                )
+            | None ->
+                processor.Post(Msg.CreateFreeMsg<_>(allIndices))
```

Comment:
We can reduce the number of duplications if we first clear the resources (by `ClArra.Free` extension) and then apply `Option.map`